src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -178,7 +178,9 @@ private[concurrent] object Promise {
    * DefaultPromises, and `linkedRootOf` is currently only designed to be called
    * by Future.flatMap.
    */
-  final class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
+  // Left non-final to enable addition of extra fields by Java/Scala converters
+  // in scala-java8-compat.
+  class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
```

Comment:
@viktorklang 
>You may ask yourself why the AtomicFieldExample is larger then the AtomicExample. This is caused by the extra memory you need to store the references + longs. AtomicFieldExample has 10 longs + 10 references. This gives us:

>10 * 8 bytes (for the longs)
10 * 4 bytes (for the references)
1 * 16 bytes (for itself)
AtomicExample has 20 refernces. This gives us:

>20 * 4 bytes (for the references)
1 * 16 bytes (for itself)
So it only pays off because we save the extra memory overhead of AtomicLong and AtomicReference itself. So to put it straight: Every Object has a fixed overhead.

from the blog:)
In netty,just using AtomicReferenceFieldUpdater.